File: src/test/java/ru/jerael/booktracker/backend/web/mapper/BookWebMapperTest.java
package ru.jerael.booktracker.backend.web.mapper;

import org.junit.jupiter.api.Test;
import ru.jerael.booktracker.backend.domain.model.Genre;
import ru.jerael.booktracker.backend.domain.model.book.Book;
import ru.jerael.booktracker.backend.domain.model.book.BookCreation;
import ru.jerael.booktracker.backend.domain.model.book.BookDetailsUpdate;
import ru.jerael.booktracker.backend.domain.model.book.BookStatus;
import ru.jerael.booktracker.backend.web.dto.book.BookCreationRequest;
import ru.jerael.booktracker.backend.web.dto.book.BookDetailsUpdateRequest;
import ru.jerael.booktracker.backend.web.dto.book.BookResponse;
import ru.jerael.booktracker.backend.web.util.LinkBuilder;
import java.time.Instant;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class BookWebMapperTest {
    private final GenreWebMapper genreWebMapper = new GenreWebMapper();
    private final AuthorWebMapper authorWebMapper = new AuthorWebMapper();
    private final PublisherWebMapper publisherWebMapper = new PublisherWebMapper();
    private final LanguageWebMapper languageWebMapper = new LanguageWebMapper();
    private final ReadingSessionWebMapper readingSessionWebMapper = new ReadingSessionWebMapper();
    private final ReadingAttemptWebMapper readingAttemptWebMapper =
        new ReadingAttemptWebMapper(readingSessionWebMapper);
    private final NoteWebMapper noteWebMapper = new NoteWebMapper();
    private final LinkBuilder linkBuilder = mock(LinkBuilder.class);
    private final BookWebMapper bookWebMapper =
        new BookWebMapper(genreWebMapper, authorWebMapper, publisherWebMapper, languageWebMapper,
            readingAttemptWebMapper, noteWebMapper, linkBuilder);

    private final UUID id = UUID.fromString("ee39af7a-a073-4473-878a-1aae34e98bb7");
    private final UUID userId = UUID.fromString("2c5781ea-1bc2-4561-a83d-26106df2526e");
    private final String title = "title";
    private final String coverFileName = "cover.jpg";
    private final Instant createdAt = Instant.ofEpochMilli(1771249699347L);
    private final Genre genre1 = new Genre(1, "action");
    private final Genre genre2 = new Genre(2, "adventure");
    private final Set<Genre> genres = Set.of(genre1, genre2);
    private final Book book = new Book(
        id,
        userId,
        title,
        coverFileName,
        createdAt,
        genres,
        Set.of(),
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        List.of(),
        List.of()
    );

    @Test
    void toResponse() {
        String coverUrl = "http://localhost:8080/api/v1/books/" + id + "/cover";
        when(linkBuilder.buildCoverUrl(id)).thenReturn(coverUrl);

        BookResponse bookResponse = bookWebMapper.toResponse(book);

        assertEquals(id, bookResponse.id());
        assertEquals(title, bookResponse.title());
        assertEquals(coverUrl, bookResponse.coverUrl());
        assertEquals(BookStatus.defaultStatus().getValue(), bookResponse.status());
        assertEquals(createdAt, bookResponse.createdAt());
        assertTrue(bookResponse.genres().containsAll(genreWebMapper.toResponses(genres)));
    }

    @Test
    void toResponses() {
        UUID id2 = UUID.fromString("31d3f5e3-7faf-4678-a3cf-4657d8875a82");
        Book book2 = new Book(
            id2,
            userId,
            "asd",
            coverFileName,
            createdAt,
            genres,
            Set.of(),
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            List.of(),
            List.of()
        );
        List<Book> books = List.of(book, book2);
        when(linkBuilder.buildCoverUrl(id)).thenReturn("url1");
        when(linkBuilder.buildCoverUrl(id2)).thenReturn("url2");

        List<BookResponse> bookResponses = bookWebMapper.toResponses(books);

        assertEquals(2, bookResponses.size());
        assertEquals(id, bookResponses.get(0).id());
        assertEquals(title, bookResponses.get(0).title());
        assertEquals(id2, bookResponses.get(1).id());
        assertEquals("asd", bookResponses.get(1).title());
        assertEquals("url1", bookResponses.get(0).coverUrl());
        assertEquals("url2", bookResponses.get(1).coverUrl());
    }

    @Test
    void toDomain_BookCreation() {
        BookCreationRequest request =
            new BookCreationRequest(" title ", " author ", "reading", Set.of(1, 2));

        BookCreation data = bookWebMapper.toDomain(request);

        assertEquals(request.title().trim(), data.title());
        assertEquals(request.author().trim(), data.author());
        assertEquals(BookStatus.fromString(request.status()), data.status());
        assertTrue(data.genreIds().containsAll(request.genreIds()));
    }

    @Test
    void toDomain_BookDetailsUpdate() {
        BookDetailsUpdateRequest request = new BookDetailsUpdateRequest(
            " title ",
            " author ",
            "reading",
            Set.of(1, 2)
        );

        BookDetailsUpdate data = bookWebMapper.toDomain(request);

        assertEquals(request.title(), data.title());
        assertEquals(request.author(), data.author());
        assertTrue(data.genreIds().containsAll(request.genreIds()));
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/web/mapper/BookWebMapperTest.kt
package ru.jerael.booktracker.backend.web.mapper

import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import ru.jerael.booktracker.backend.domain.model.book.BookStatus
import ru.jerael.booktracker.backend.factory.book.BookDomainFactory
import ru.jerael.booktracker.backend.factory.book.BookWebFactory
import ru.jerael.booktracker.backend.web.util.LinkBuilder

@ExtendWith(MockKExtension::class)
class BookWebMapperTest {
    private val genreWebMapper = GenreWebMapper()
    private val authorWebMapper = AuthorWebMapper()
    private val publisherWebMapper = PublisherWebMapper()
    private val languageWebMapper = LanguageWebMapper()
    private val readingSessionWebMapper = ReadingSessionWebMapper()
    private val readingAttemptWebMapper = ReadingAttemptWebMapper(readingSessionWebMapper)
    private val noteWebMapper = NoteWebMapper()
    private val linkBuilder = mockk<LinkBuilder>()
    
    @InjectMockKs
    private lateinit var mapper: BookWebMapper
    
    @Test
    fun `toResponse should map Book to BookResponse`() {
        val book = BookDomainFactory.createBook(coverFileName = "cover.jpg")
        val coverUrl = "http://localhost:8080/api/v1/books/${book.id}/cover"
        every { linkBuilder.buildCoverUrl(book.id) } returns coverUrl
        
        val bookResponse = mapper.toResponse(book)
        
        with(bookResponse) {
            assertEquals(book.id, id)
            assertEquals(book.title, title)
            assertEquals(coverUrl, this.coverUrl)
            assertEquals(book.status().value, status)
            assertEquals(book.createdAt, createdAt)
            
            val expectedGenres = genreWebMapper.toResponses(book.genres)
            assertEquals(expectedGenres, genres)
            
            val expectedAuthors = book.authors.map { authorWebMapper.toResponse(it) }.toSet()
            assertEquals(expectedAuthors, authors)
            
            assertEquals(book.description, description)
            
            val expectedPublisher = publisherWebMapper.toResponse(book.publisher)
            assertEquals(expectedPublisher, publisher)
            
            val expectedLanguage = languageWebMapper.toResponse(book.language)
            assertEquals(expectedLanguage, language)
            
            assertEquals(book.publishedOn, publishedOn)
            assertEquals(book.totalPages, totalPages)
            assertEquals(book.isbn10, isbn10)
            assertEquals(book.isbn13, isbn13)
            
            val expectedReadingAttempts = book.attempts.map { readingAttemptWebMapper.toResponse(it) }
            assertEquals(expectedReadingAttempts, attempts)
            
            val expectedNotes = book.notes.map { noteWebMapper.toResponse(it) }
            assertEquals(expectedNotes, notes)
        }
    }
    
    @Test
    fun `toResponses should map Books to BookResponses`() {
        val book = BookDomainFactory.createBook(coverFileName = "cover.jpg")
        val book2 = BookDomainFactory.createBook(userId = book.userId, title = "asd", coverFileName = "cover.jpg")
        every { linkBuilder.buildCoverUrl(book.id) } returns "url1"
        every { linkBuilder.buildCoverUrl(book2.id) } returns "url2"
        
        val bookResponses = mapper.toResponses(listOf(book, book2))
        
        with(bookResponses) {
            assertEquals(2, size)
            
            assertEquals(book.id, this[0].id)
            assertEquals(book.title, this[0].title)
            assertEquals("url1", this[0].coverUrl)
            
            assertEquals(book2.id, this[1].id)
            assertEquals(book2.title, this[1].title)
            assertEquals("url2", this[1].coverUrl)
        }
    }
    
    @Test
    fun `toDomain should map BookCreationRequest with userId to BookCreation`() {
        val book = BookDomainFactory.createBook()
        val request = BookWebFactory.createBookCreationRequest(
            isbn10 = "123-456-789-0",
            isbn13 = "123-456-789-012-3"
        )
        
        val data = mapper.toDomain(request, book.userId)
        
        with(data) {
            assertEquals(book.userId, userId)
            assertEquals(request.title.trim(), title)
            assertEquals(BookStatus.fromString(request.status), status)
            assertEquals(request.genreIds, genreIds)
            assertEquals(request.authorNames, authorNames)
            assertEquals(request.description, description)
            assertEquals(request.publisherName, publisherName)
            assertEquals(request.languageCode, languageCode)
            assertEquals(request.publishedOn, publishedOn)
            assertEquals(request.totalPages, totalPages)
            assertEquals("1234567890", isbn10)
            assertEquals("1234567890123", isbn13)
        }
    }
    
    @Test
    fun `toDomain should map BookDetailsUpdateRequest with bookId and userId to BookDetailsUpdate`() {
        val book = BookDomainFactory.createBook()
        val request = BookWebFactory.createBookDetailsUpdateRequest(
            isbn10 = "123-456-789-0",
            isbn13 = "123-456-789-012-3"
        )
        
        val data = mapper.toDomain(request, book.id, book.userId)
        
        with(data) {
            assertEquals(book.userId, userId)
            assertEquals(request.title, title)
            assertEquals(BookStatus.fromString(request.status), status)
            assertEquals(request.genreIds, genreIds)
            assertEquals(request.authorNames, authorNames)
            assertEquals(request.description, description)
            assertEquals(request.publisherName, publisherName)
            assertEquals(request.languageCode, languageCode)
            assertThat(publishedOn as Int?).isEqualTo(request.publishedOn as Int?)
            assertThat(totalPages as Int?).isEqualTo(request.totalPages as Int?)
            assertEquals("1234567890", isbn10)
            assertEquals("1234567890123", isbn13)
        }
    }
    
    @Test
    fun `toResponse should map BookMetadata to BookMetadataResponse`() {
        val metadata = BookDomainFactory.createBookMetadata()
        
        val result = mapper.toResponse(metadata)
        
        with(result) {
            assertThat(title).isEqualTo(metadata.title)
            assertThat(coverUrl).isEqualTo(metadata.cover)
            
            val expectedGenres = genreWebMapper.toResponses(metadata.genres)
            assertThat(genres).isEqualTo(expectedGenres)
            
            val expectedAuthors = metadata.authors.map { authorWebMapper.toResponse(it) }.toSet()
            assertThat(authors).isEqualTo(expectedAuthors)
            
            assertThat(description).isEqualTo(metadata.description)
            
            val expectedPublisher = publisherWebMapper.toResponse(metadata.publisher)
            assertThat(publisher).isEqualTo(expectedPublisher)
            
            val expectedLanguage = languageWebMapper.toResponse(metadata.language)
            assertThat(language).isEqualTo(expectedLanguage)
            
            assertThat(publishedOn).isEqualTo(metadata.publishedOn)
            assertThat(totalPages).isEqualTo(metadata.totalPages)
            assertThat(isbn10).isEqualTo(metadata.isbn10)
            assertThat(isbn13).isEqualTo(metadata.isbn13)
        }
    }
}
